algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i

```diff
@@ -124,46 +124,28 @@ template <typename algorithmFPType, CpuType cpu>
 services::Status KernelImplRBF<defaultDense, algorithmFPType, cpu>::computeInternalMatrixMatrix(const NumericTable * a1, const NumericTable * a2,
                                                                                                 NumericTable * r, const ParameterBase * par)
 {
-    //prepareData
+    SafeStatus safeStat;
+
     const size_t nVectors1 = a1->getNumberOfRows();
     const size_t nVectors2 = a2->getNumberOfRows();
     const size_t nFeatures = a1->getNumberOfColumns();
 
-    ReadRows<algorithmFPType, cpu> mtA1(*const_cast<NumericTable *>(a1), 0, nVectors1);
-    DAAL_CHECK_BLOCK_STATUS(mtA1);
-    const algorithmFPType * dataA1 = mtA1.get();
-
-    WriteOnlyRows<algorithmFPType, cpu> mtR(r, 0, nVectors1);
-    DAAL_CHECK_BLOCK_STATUS(mtR);
-    algorithmFPType * dataR = mtR.get();
-
-    //compute
     const Parameter * rbfPar    = static_cast<const Parameter *>(par);
     const algorithmFPType coeff = (algorithmFPType)(-0.5 / (rbfPar->sigma * rbfPar->sigma));
 
-    bool isInParallel = is_in_parallel();
     if (a1 != a2)
     {
-        ReadRows<algorithmFPType, cpu> mtA2(*const_cast<NumericTable *>(a2), 0, nVectors2);
-        DAAL_CHECK_BLOCK_STATUS(mtA2);
-        const algorithmFPType * dataA2 = mtA2.get();
-
-        char trans, notrans;
+        char trans = 'T', notrans = 'N';
         algorithmFPType zero = 0.0, negTwo = -2.0;
-        trans   = 'T';
-        notrans = 'N';
-        if (isInParallel)
-        {
-            Blas<algorithmFPType, cpu>::xxgemm(&trans, &notrans, (DAAL_INT *)&nVectors2, (DAAL_INT *)&nVectors1, (DAAL_INT *)&nFeatures, &negTwo,
-                                               (algorithmFPType *)dataA2, (DAAL_INT *)&nFeatures, (algorithmFPType *)dataA1, (DAAL_INT *)&nFeatures,
-                                               &zero, dataR, (DAAL_INT *)&nVectors2);
-        }
-        else
-        {
-            Blas<algorithmFPType, cpu>::xgemm(&trans, &notrans, (DAAL_INT *)&nVectors2, (DAAL_INT *)&nVectors1, (DAAL_INT *)&nFeatures, &negTwo,
-                                              (algorithmFPType *)dataA2, (DAAL_INT *)&nFeatures, (algorithmFPType *)dataA1, (DAAL_INT *)&nFeatures,
-                                              &zero, dataR, (DAAL_INT *)&nVectors2);
-        }
+
+        services::Status retStat = Blas<algorithmFPType, cpu>::xgemm_blocked(&trans, &notrans, (DAAL_INT *)&nVectors2, (DAAL_INT *)&nVectors1,
+                                                                             (DAAL_INT *)&nFeatures, &negTwo, a2, (DAAL_INT *)&nFeatures, a1,
+                                                                             (DAAL_INT *)&nFeatures, &zero, r, (DAAL_INT *)&nVectors2);
+        if (!retStat) return retStat;
```

Comment:
In case of RBF kernel, use of xgemm_blocked is not the best option.
In your variant of code you have additional passes through a1 and a2 to compute sqrDataA1 and sqrDataA2.
In my opinion, it is better to embed those computations into threader_for which does matrix multiplication a1 x a2 in cache-friendly manner.

Reply:
I modified blocked gemm. So now gemm and sqrDataA1 calculated in the same threader_for. But I couldn't do the same for sqrDataA2 because of parallelism